In [ ]:
# Correct import for pandas in Jupyter
import pandas as pd


In [13]:
df = pd.read_csv('../data/raw/Samsung.csv')
df.head(30)

Year Quarter      Product Model 5G Capability  Units Sold   Revenue ($)  \
0   2019      Q1         Galaxy S10            No       26396  4.212951e+06   
1   2019      Q1      Galaxy Note10            No       25671  7.240266e+06   
2   2019      Q1         Galaxy S20            No       16573  2.560833e+07   
3   2019      Q1      Galaxy Note20            No        7177  2.198442e+07   
4   2019      Q1         Galaxy S21            No       45633  1.634244e+07   
5   2019      Q1      Galaxy A32 5G           Yes       15912  1.717833e+07   
6   2019      Q1      Galaxy A52 5G           Yes        7231  4.863981e+07   
7   2019      Q1      Galaxy A73 5G           Yes       58711  7.017578e+07   
8   2019      Q1  Galaxy Z Fold2 5G           Yes       40641  5.195956e+07   
9   2019      Q1  Galaxy Z Flip3 5G           Yes       38119  4.989954e+07   
10  2019      Q1      Galaxy S22 5G           Yes       36255  1.209167e+07   
11  2019      Q1  Galaxy Z Fold3 5G           Yes       19643  3.602950e+07   
12  2019      Q1      Galaxy A14 5G           Yes       12612  5.716155e+07   
13  2019      Q1      Galaxy S23 5G           Yes       22722  4.531095e+07   
14  2019      Q1  Galaxy Z Flip5 5G           Yes        8453  4.693728e+07   
15  2019      Q2         Galaxy S10            No       19160  6.660065e+06   
16  2019      Q2      Galaxy Note10            No       38253  5.319594e+07   
17  2019      Q2         Galaxy S20            No       17404  1.495021e+07   
18  2019      Q2      Galaxy Note20            No       27841  9.787427e+06   
19  2019      Q2         Galaxy S21            No       10047  1.069152e+07   
20  2019      Q2      Galaxy A32 5G           Yes       25054  4.049594e+07   
21  2019      Q2      Galaxy A52 5G           Yes       37263  2.976331e+07   
22  2019      Q2      Galaxy A73 5G           Yes       39191  1.963191e+07   
23  2019      Q2  Galaxy Z Fold2 5G           Yes       11320  3.667825e+07   
24  2019      Q2  Galaxy Z Flip3 5G           Yes       31105  1.281874e+07   
25  2019      Q2      Galaxy S22 5G           Yes       24185  5.219297e+07   
26  2019      Q2  Galaxy Z Fold3 5G           Yes        8422  1.129249e+07   
27  2019      Q2      Galaxy A14 5G           Yes       56397  1.839088e+07   
28  2019      Q2      Galaxy S23 5G           Yes       36613  6.307077e+07   
29  2019      Q2  Galaxy Z Flip5 5G           Yes       12030  3.354957e+07   

    Market Share (%)  Regional 5G Coverage (%)  5G Subscribers (millions)  \
0               1.04                     57.36                      39.55   
1               2.82                     85.80                      42.58   
2              -0.03                     47.02                       3.78   
3               0.84                     25.70                      23.41   
4               2.36                     89.13                      44.43   
5               5.41                     59.12                      12.14   
6               5.90                     52.94                      44.36   
7               4.92                     79.95                      49.70   
8               2.64                     44.77                      31.27   
9               6.84                     51.21                      38.33   
10              4.66                     56.50                       9.77   
11              5.67                     76.62                      22.84   
12              3.24                     64.48                      14.98   
13              2.88                     87.12                      19.04   
14              3.48                     72.73                      23.32   
15              1.56                     67.69                      14.60   
16             -0.05                     76.73                      31.96   
17              2.51                     38.86                      36.38   
18              3.96                     75.22                      46.41   
19            

In [14]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 12 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Year                       1000 non-null   int64  
 1   Quarter                    1000 non-null   object 
 2   Product Model              1000 non-null   object 
 3   5G Capability              1000 non-null   object 
 4   Units Sold                 1000 non-null   int64  
 5   Revenue ($)                1000 non-null   float64
 6   Market Share (%)           1000 non-null   float64
 7   Regional 5G Coverage (%)   1000 non-null   float64
 8   5G Subscribers (millions)  1000 non-null   float64
 9   Avg 5G Speed (Mbps)        1000 non-null   float64
 10  Preference for 5G (%)      1000 non-null   float64
 11  Region                     1000 non-null   object 
dtypes: float64(6), int64(2), object(4)
memory usage: 93.9+ KB


In [15]:
df.describe()

Year    Units Sold   Revenue ($)  Market Share (%)  \
count  1000.000000   1000.000000  1.000000e+03       1000.000000   
mean   2021.456000  32842.990000  3.019733e+07          3.723570   
std       1.700281  16039.770572  1.837941e+07          1.991108   
min    2019.000000   5309.000000  2.987436e+06         -0.490000   
25%    2020.000000  19327.250000  1.460749e+07          2.635000   
50%    2021.000000  33689.000000  2.801200e+07          3.760000   
75%    2023.000000  43911.000000  4.180391e+07          5.282500   
max    2024.000000  64883.000000  8.426494e+07          6.950000   

       Regional 5G Coverage (%)  5G Subscribers (millions)  \
count               1000.000000                1000.000000   
mean                  66.889720                  30.152080   
std                   19.254095                  14.537781   
min                   25.340000                  -0.890000   
25%                   50.400000                  18.412500   
50%                   67.050000                  29.915000   
75%                   83.210000                  44.360000   
max                  103.920000                  54.940000   

       Avg 5G Speed (Mbps)  Preference for 5G (%)  
count          1000.000000             1000.00000  
mean            179.225560               67.14291  
std              70.470934               15.75925  
min              50.370000               37.14000  
25%             120.410000               53.26750  
50%             177.390000               66.96000  
75%             238.860000               80.99000  
max             298.700000               94.84000

In [16]:
df.isna().sum()

Year                         0
Quarter                      0
Product Model                0
5G Capability                0
Units Sold                   0
Revenue ($)                  0
Market Share (%)             0
Regional 5G Coverage (%)     0
5G Subscribers (millions)    0
Avg 5G Speed (Mbps)          0
Preference for 5G (%)        0
Region                       0
dtype: int64

In [17]:
df.duplicated().sum()

np.int64(640)

In [18]:
df.nunique()

Year                           6
Quarter                        4
Product Model                 15
5G Capability                  2
Units Sold                   354
Revenue ($)                  360
Market Share (%)             274
Regional 5G Coverage (%)     344
5G Subscribers (millions)    350
Avg 5G Speed (Mbps)          358
Preference for 5G (%)        347
Region                         5
dtype: int64

## what is “not sensible"?

Based on outputs, the following values are suspicious:

1. Negative market share 
2. Negative 5G subscribers
3. 5G coverage above 100%
4. 640 duplicate rows

***Everything else seems normal***

In [23]:
df.drop_duplicates(inplace=True)

In [26]:
df.info()
df.describe()


<class 'pandas.core.frame.DataFrame'>
Index: 360 entries, 0 to 359
Data columns (total 12 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Year                       360 non-null    int64  
 1   Quarter                    360 non-null    object 
 2   Product Model              360 non-null    object 
 3   5G Capability              360 non-null    object 
 4   Units Sold                 360 non-null    int64  
 5   Revenue ($)                360 non-null    float64
 6   Market Share (%)           360 non-null    float64
 7   Regional 5G Coverage (%)   360 non-null    float64
 8   5G Subscribers (millions)  360 non-null    float64
 9   Avg 5G Speed (Mbps)        360 non-null    float64
 10  Preference for 5G (%)      360 non-null    float64
 11  Region                     360 non-null    object 
dtypes: float64(6), int64(2), object(4)
memory usage: 36.6+ KB


Year    Units Sold   Revenue ($)  Market Share (%)  \
count   360.000000    360.000000  3.600000e+02        360.000000   
mean   2021.500000  32646.219444  2.990730e+07          3.716167   
std       1.710202  16313.428687  1.848561e+07          1.985063   
min    2019.000000   5309.000000  2.987436e+06         -0.490000   
25%    2020.000000  18954.500000  1.445305e+07          2.617500   
50%    2021.500000  33018.000000  2.784645e+07          3.710000   
75%    2023.000000  44533.750000  4.160702e+07          5.280000   
max    2024.000000  64883.000000  8.426494e+07          6.950000   

       Regional 5G Coverage (%)  5G Subscribers (millions)  \
count                360.000000                 360.000000   
mean                  66.885194                  30.137194   
std                   19.430997                  14.551242   
min                   25.340000                  -0.890000   
25%                   50.332500                  18.517500   
50%                   67.000000                  29.915000   
75%                   83.210000                  44.267500   
max                  103.920000                  54.940000   

       Avg 5G Speed (Mbps)  Preference for 5G (%)  
count           360.000000             360.000000  
mean            179.972944              67.291778  
std              70.657616              15.590697  
min              50.370000              37.140000  
25%             122.707500              53.537500  
50%             178.735000              67.475000  
75%             239.805000              80.832500  
max             298.700000              94.840000

In [ ]:
#Create flags for bad values before fixing
df["Negative_MarketShare_Flag"] = df["Market Share (%)"] < 0
df["Negative_5GSubscribers_Flag"] = df["5G Subscribers (millions)"] < 0
df["Coverage_Above_100_Flag"] = df["Regional 5G Coverage (%)"] > 100


In [ ]:
#Fix negative and out-of-bounds values
df.loc[df["Market Share (%)"] < 0, "Market Share (%)"] = 0
df.loc[df["5G Subscribers (millions)"] < 0, "5G Subscribers (millions)"] = 0
df["Regional 5G Coverage (%)"] = df["Regional 5G Coverage (%)"].clip(0, 100)

In [49]:
#Convert Quarter (e.g., "Q1") to month (Q1→01, Q2→04, Q3→07, Q4→10)
quarter_to_month = {"Q1": "01", "Q2": "04", "Q3": "07", "Q4": "10"}
df["Month"] = df["Quarter"].map(quarter_to_month)
df["Date"] = pd.to_datetime(df["Year"].astype(str) + "-" + df["Month"] + "-01")



In [48]:
#Create a Quarter_Index for each Product Model
#It let us track the number of quarters since a product model was launched
# We can know how long a product model has been on the market
# This can be useful for analyzing trends related to product lifecycle
# For example, we can see if market share tends to decline after a certain number of quarters
df = df.sort_values(by="Date")
df["Quarter_Index"] = df.groupby("Product Model").cumcount() + 1


In [55]:
df["Is_5G"] = df["5G Capability"].apply(lambda x: 1 if x == 'Yes' else 0)
df.head(30)
df["5G_Rediness_Score"] = (((df["Avg 5G Speed (Mbps)"] / df["Avg 5G Speed (Mbps)"].max()) * 0.5 )+ 
                          (df["Regional 5G Coverage (%)"] / 100 * 0.5))
df.head(30)

Year Quarter      Product Model 5G Capability  Units Sold   Revenue ($)  \
0   2019      Q1         Galaxy S10            No       26396  4.212951e+06   
3   2019      Q1      Galaxy Note20            No        7177  2.198442e+07   
4   2019      Q1         Galaxy S21            No       45633  1.634244e+07   
7   2019      Q1      Galaxy A73 5G           Yes       58711  7.017578e+07   
1   2019      Q1      Galaxy Note10            No       25671  7.240266e+06   
5   2019      Q1      Galaxy A32 5G           Yes       15912  1.717833e+07   
6   2019      Q1      Galaxy A52 5G           Yes        7231  4.863981e+07   
2   2019      Q1         Galaxy S20            No       16573  2.560833e+07   
8   2019      Q1  Galaxy Z Fold2 5G           Yes       40641  5.195956e+07   
9   2019      Q1  Galaxy Z Flip3 5G           Yes       38119  4.989954e+07   
13  2019      Q1      Galaxy S23 5G           Yes       22722  4.531095e+07   
14  2019      Q1  Galaxy Z Flip5 5G           Yes        8453  4.693728e+07   
10  2019      Q1      Galaxy S22 5G           Yes       36255  1.209167e+07   
11  2019      Q1  Galaxy Z Fold3 5G           Yes       19643  3.602950e+07   
12  2019      Q1      Galaxy A14 5G           Yes       12612  5.716155e+07   
17  2019      Q2         Galaxy S20            No       17404  1.495021e+07   
29  2019      Q2  Galaxy Z Flip5 5G           Yes       12030  3.354957e+07   
19  2019      Q2         Galaxy S21            No       10047  1.069152e+07   
27  2019      Q2      Galaxy A14 5G           Yes       56397  1.839088e+07   
20  2019      Q2      Galaxy A32 5G           Yes       25054  4.049594e+07   
15  2019      Q2         Galaxy S10            No       19160  6.660065e+06   
21  2019      Q2      Galaxy A52 5G           Yes       37263  2.976331e+07   
18  2019      Q2      Galaxy Note20            No       27841  9.787427e+06   
23  2019      Q2  Galaxy Z Fold2 5G           Yes       11320  3.667825e+07   
28  2019      Q2      Galaxy S23 5G           Yes       36613  6.307077e+07   
24  2019      Q2  Galaxy Z Flip3 5G           Yes       31105  1.281874e+07   
25  2019      Q2      Galaxy S22 5G           Yes       24185  5.219297e+07   
16  2019      Q2      Galaxy Note10            No       38253  5.319594e+07   
26  2019      Q2  Galaxy Z Fold3 5G           Yes        8422  1.129249e+07   
22  2019      Q2      Galaxy A73 5G           Yes       39191  1.963191e+07   

    Market Share (%)  Regional 5G Coverage (%)  5G Subscribers (millions)  \
0               1.04                     57.36                      39.55   
3               0.84                     25.70                      23.41   
4               2.36                     89.13                      44.43   
7               4.92                     79.95                      49.70   
1               2.82                     85.80                      42.58   
5               5.41                     59.12                      12.14   
6               5.90                     52.94                      44.36   
2               0.00                     47.02                       3.78   
8               2.64                     44.77                      31.27   
9               6.84                     51.21                      38.33   
13              2.88                     87.12                      19.04   
14              3.48                     72.73                      23.32   
10              4.66                     56.50                       9.77   
11              5.67                     76.62                      22.84   
12              3.24                     64.48                      14.98   
17              2.51                     38.86                      36.38   
29              5.82                     48.62                      36.32   
19              1.56                     45.75                      15.36   
27              6.11                     70.96                      23.80   
20            

In [2]:
import matplotlib.pyplot as plt
